File: services/iceberg-catalog-service/src/domain/schema_strict.rs
//! Schema strict-mode enforcement.
//!
//! `Iceberg tables.md` § "Notable differences" / "Automatic schema
//! evolution" pins the rule:
//!
//! > Iceberg is strict about the schema when writing to an existing
//! > Iceberg table. Any change in schema needs to be made explicitly
//! > via an `ALTER TABLE` command.
//!
//! The catalog enforces this invariant in two places:
//!
//!   1. [`diff_schemas`] computes the structural difference between
//!      the table's current schema and the schema a writer is about
//!      to commit. When the diff is non-empty the commit is rejected
//!      with `422 SCHEMA_INCOMPATIBLE_REQUIRES_ALTER`.
//!   2. The dedicated `POST /alter-schema` endpoint (in
//!      `handlers::rest_catalog::tables`) accepts an explicit list of
//!      schema mutations and bumps `schema-id`.
//!
//! This module is stateless — it just owns the diff algorithm and the
//! response envelope so handlers stay focused on HTTP concerns.

use serde::Serialize;
use serde_json::Value;
use std::collections::BTreeMap;

/// Granular description of one schema-level change. Used by the
/// `422` response envelope so clients (especially the
/// pipeline-authoring UI's "generate ALTER TABLE" CTA) can build the
/// migration without re-running the diff client-side.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum SchemaDelta {
    AddedColumn {
        name: String,
        column_type: String,
    },
    DroppedColumn {
        name: String,
        column_type: String,
    },
    ChangedColumnType {
        name: String,
        from: String,
        to: String,
    },
    ChangedColumnRequired {
        name: String,
        from: bool,
        to: bool,
    },
}

/// Full diff envelope. The list is empty when the schemas are
/// equivalent.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct SchemaDiff {
    pub deltas: Vec<SchemaDelta>,
}

impl SchemaDiff {
    pub fn is_compatible(&self) -> bool {
        self.deltas.is_empty()
    }

    pub fn rendered(&self) -> String {
        self.deltas
            .iter()
            .map(|d| match d {
                SchemaDelta::AddedColumn { name, column_type } => {
                    format!("+{name}:{column_type}")
                }
                SchemaDelta::DroppedColumn { name, column_type } => {
                    format!("-{name}:{column_type}")
                }
                SchemaDelta::ChangedColumnType { name, from, to } => {
                    format!("~{name}:{from}→{to}")
                }
                SchemaDelta::ChangedColumnRequired { name, from, to } => {
                    format!("?{name}:{from}→{to}")
                }
            })
            .collect::<Vec<_>>()
            .join(", ")
    }
}

/// Compare two Iceberg schemas (the JSON shape that lives in
/// `iceberg_tables.schema_json`). Returns the deltas that the caller
/// must explicitly apply via the alter-schema endpoint before the
/// commit can land.
pub fn diff_schemas(current: &Value, attempted: &Value) -> SchemaDiff {
    let current_fields = extract_fields(current);
    let attempted_fields = extract_fields(attempted);

    let mut deltas = Vec::new();
    for (name, attr) in attempted_fields.iter() {
        match current_fields.get(name) {
            None => deltas.push(SchemaDelta::AddedColumn {
                name: name.clone(),
                column_type: attr.column_type.clone(),
            }),
            Some(curr) if curr.column_type != attr.column_type => {
                deltas.push(SchemaDelta::ChangedColumnType {
                    name: name.clone(),
                    from: curr.column_type.clone(),
                    to: attr.column_type.clone(),
                });
            }
            Some(curr) if curr.required != attr.required => {
                deltas.push(SchemaDelta::ChangedColumnRequired {
                    name: name.clone(),
                    from: curr.required,
                    to: attr.required,
                });
            }
            _ => {}
        }
    }
    for (name, attr) in current_fields.iter() {
        if !attempted_fields.contains_key(name) {
            deltas.push(SchemaDelta::DroppedColumn {
                name: name.clone(),
                column_type: attr.column_type.clone(),
            });
        }
    }
    SchemaDiff { deltas }
}

#[derive(Debug, Clone)]
struct FieldAttrs {
    column_type: String,
    required: bool,
}

/// Extract the `{ name, type, required }` triples from an Iceberg
/// schema JSON. The function tolerates both v1 (where `type` may be a
/// nested object) and v2 (string scalars).
fn extract_fields(schema: &Value) -> BTreeMap<String, FieldAttrs> {
    let fields = schema
        .get("fields")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    fields
        .into_iter()
        .filter_map(|field| {
            let name = field.get("name")?.as_str()?.to_string();
            let column_type = match field.get("type")? {
                Value::String(s) => s.clone(),
                other => serde_json::to_string(other).unwrap_or_default(),
            };
            let required = field
                .get("required")
                .and_then(Value::as_bool)
                .unwrap_or(false);
            Some((
                name,
                FieldAttrs {
                    column_type,
                    required,
                },
            ))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn schema_with_fields(fields: Vec<Value>) -> Value {
        json!({ "schema-id": 0, "type": "struct", "fields": fields })
    }

    #[test]
    fn identical_schemas_are_compatible() {
        let s = schema_with_fields(vec![
            json!({ "id": 1, "name": "id", "required": true, "type": "long" }),
            json!({ "id": 2, "name": "name", "required": false, "type": "string" }),
        ]);
        assert!(diff_schemas(&s, &s).is_compatible());
    }

    #[test]
    fn added_column_is_detected() {
        let current = schema_with_fields(vec![json!({
            "id": 1, "name": "id", "required": true, "type": "long"
        })]);
        let attempted = schema_with_fields(vec![
            json!({ "id": 1, "name": "id", "required": true, "type": "long" }),
            json!({ "id": 2, "name": "added", "required": false, "type": "string" }),
        ]);
        let diff = diff_schemas(&current, &attempted);
        assert!(!diff.is_compatible());
        assert!(diff.deltas.iter().any(|d| matches!(
            d,
            SchemaDelta::AddedColumn { name, .. } if name == "added"
        )));
    }

    #[test]
    fn dropped_column_is_detected() {
        let current = schema_with_fields(vec![
            json!({ "id": 1, "name": "id", "required": true, "type": "long" }),
            json!({ "id": 2, "name": "removed", "required": false, "type": "string" }),
        ]);
        let attempted = schema_with_fields(vec![json!({
            "id": 1, "name": "id", "required": true, "type": "long"
        })]);
        let diff = diff_schemas(&current, &attempted);
        assert!(diff.deltas.iter().any(|d| matches!(
            d,
            SchemaDelta::DroppedColumn { name, .. } if name == "removed"
        )));
    }

    #[test]
    fn changed_column_type_is_detected() {
        let current = schema_with_fields(vec![json!({
            "id": 1, "name": "id", "required": true, "type": "long"
        })]);
        let attempted = schema_with_fields(vec![json!({
            "id": 1, "name": "id", "required": true, "type": "string"
        })]);
        let diff = diff_schemas(&current, &attempted);
        assert!(diff.deltas.iter().any(|d| matches!(
            d,
            SchemaDelta::ChangedColumnType { name, from, to }
                if name == "id" && from == "long" && to == "string"
        )));
    }
}

File: services/iceberg-catalog-service/internal/domain/schema_strict.go
// Package domain — schema strict-mode enforcement.
//
// `Iceberg tables.md` § "Notable differences" / "Automatic schema
// evolution" pins the rule:
//
//	> Iceberg is strict about the schema when writing to an existing
//	> Iceberg table. Any change in schema needs to be made explicitly
//	> via an ALTER TABLE command.
//
// The catalog enforces this invariant in two places:
//
//  1. DiffSchemas computes the structural difference between the
//     table's current schema and the schema a writer is about to
//     commit. When the diff is non-empty the commit is rejected
//     with 422 SCHEMA_INCOMPATIBLE_REQUIRES_ALTER.
//  2. The dedicated POST /alter-schema endpoint accepts an explicit
//     list of schema mutations and bumps schema-id.
//
// This file is a 1:1 port of services/iceberg-catalog-service/src/domain/
// schema_strict.rs.
package domain

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// Schema-delta kind tags. Kebab-case to match Rust's
// #[serde(tag = "kind", rename_all = "kebab-case")] enum encoding.
const (
	DeltaAddedColumn           = "added-column"
	DeltaDroppedColumn         = "dropped-column"
	DeltaChangedColumnType     = "changed-column-type"
	DeltaChangedColumnRequired = "changed-column-required"
)

// SchemaDelta is a granular description of one schema-level change.
// Used by the 422 response envelope so clients (especially the
// pipeline-authoring UI's "generate ALTER TABLE" CTA) can build the
// migration without re-running the diff client-side.
//
// The JSON shape is the flat-tagged form Rust serde produces:
//
//	added-column / dropped-column → {"kind": ..., "name": ..., "column_type": ...}
//	changed-column-type           → {"kind": ..., "name": ..., "from": "long", "to": "string"}
//	changed-column-required       → {"kind": ..., "name": ..., "from": true, "to": false}
//
// `From`/`To` are kept as `any` because the variant determines the
// underlying type (string for type-changes, bool for required-changes).
type SchemaDelta struct {
	Kind       string `json:"kind"`
	Name       string `json:"name"`
	ColumnType string `json:"column_type,omitempty"`
	From       any    `json:"from,omitempty"`
	To         any    `json:"to,omitempty"`
}

// SchemaDiff is the full diff envelope. The list is empty when the
// schemas are equivalent.
type SchemaDiff struct {
	Deltas []SchemaDelta `json:"deltas"`
}

// IsCompatible reports whether the diff has no deltas.
func (d SchemaDiff) IsCompatible() bool { return len(d.Deltas) == 0 }

// Rendered formats the diff as a comma-separated human-readable list
// matching Rust's `SchemaDiff::rendered` output.
func (d SchemaDiff) Rendered() string {
	parts := make([]string, 0, len(d.Deltas))
	for _, delta := range d.Deltas {
		switch delta.Kind {
		case DeltaAddedColumn:
			parts = append(parts, fmt.Sprintf("+%s:%s", delta.Name, delta.ColumnType))
		case DeltaDroppedColumn:
			parts = append(parts, fmt.Sprintf("-%s:%s", delta.Name, delta.ColumnType))
		case DeltaChangedColumnType:
			parts = append(parts, fmt.Sprintf("~%s:%v→%v", delta.Name, delta.From, delta.To))
		case DeltaChangedColumnRequired:
			parts = append(parts, fmt.Sprintf("?%s:%v→%v", delta.Name, delta.From, delta.To))
		}
	}
	return strings.Join(parts, ", ")
}

type fieldAttrs struct {
	columnType string
	required   bool
}

// DiffSchemas compares two Iceberg schemas (the JSON shape that lives
// in iceberg_tables.schema_json) and returns the deltas the caller
// must explicitly apply via the alter-schema endpoint before the
// commit can land.
//
// `current` and `attempted` are raw JSON values; either may be nil
// (treated as the empty schema). The function tolerates both v1
// (where `type` may be a nested object) and v2 (string scalars) per
// the Iceberg spec.
func DiffSchemas(current, attempted json.RawMessage) SchemaDiff {
	currentFields := extractFields(current)
	attemptedFields := extractFields(attempted)

	var deltas []SchemaDelta

	// Walk attempted fields in alphabetical order to match Rust's
	// BTreeMap iteration (so the diff list is deterministic and
	// byte-exact across both implementations).
	attemptedNames := sortedKeys(attemptedFields)
	for _, name := range attemptedNames {
		attr := attemptedFields[name]
		curr, ok := currentFields[name]
		switch {
		case !ok:
			deltas = append(deltas, SchemaDelta{
				Kind:       DeltaAddedColumn,
				Name:       name,
				ColumnType: attr.columnType,
			})
		case curr.columnType != attr.columnType:
			deltas = append(deltas, SchemaDelta{
				Kind: DeltaChangedColumnType,
				Name: name,
				From: curr.columnType,
				To:   attr.columnType,
			})
		case curr.required != attr.required:
			deltas = append(deltas, SchemaDelta{
				Kind: DeltaChangedColumnRequired,
				Name: name,
				From: curr.required,
				To:   attr.required,
			})
		}
	}

	for _, name := range sortedKeys(currentFields) {
		if _, ok := attemptedFields[name]; ok {
			continue
		}
		deltas = append(deltas, SchemaDelta{
			Kind:       DeltaDroppedColumn,
			Name:       name,
			ColumnType: currentFields[name].columnType,
		})
	}

	if deltas == nil {
		deltas = []SchemaDelta{}
	}
	return SchemaDiff{Deltas: deltas}
}

// extractFields pulls the {name, type, required} triples from an
// Iceberg schema JSON. Tolerates both v1 (`type` as a nested object)
// and v2 (string scalars).
func extractFields(schema json.RawMessage) map[string]fieldAttrs {
	out := make(map[string]fieldAttrs)
	if len(schema) == 0 {
		return out
	}
	var doc struct {
		Fields []json.RawMessage `json:"fields"`
	}
	if err := json.Unmarshal(schema, &doc); err != nil {
		return out
	}
	for _, raw := range doc.Fields {
		var field map[string]json.RawMessage
		if err := json.Unmarshal(raw, &field); err != nil {
			continue
		}
		nameRaw, ok := field["name"]
		if !ok {
			continue
		}
		var name string
		if err := json.Unmarshal(nameRaw, &name); err != nil {
			continue
		}
		typeRaw, ok := field["type"]
		if !ok {
			continue
		}
		columnType := decodeColumnType(typeRaw)
		required := false
		if reqRaw, ok := field["required"]; ok {
			_ = json.Unmarshal(reqRaw, &required)
		}
		out[name] = fieldAttrs{columnType: columnType, required: required}
	}
	return out
}

// decodeColumnType mirrors Rust's serde match on Value::String vs
// other shapes: scalar strings round-trip as themselves; complex
// types (struct/list/map) round-trip through a JSON re-encode so the
// diff treats `{"type":"list","element":"long"}` and
// `{"type":"list","element":"string"}` as different column types.
func decodeColumnType(raw json.RawMessage) string {
	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		return asString
	}
	out, err := json.Marshal(raw)
	if err != nil {
		return ""
	}
	return string(out)
}

func sortedKeys(m map[string]fieldAttrs) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// SchemaIncompatibleError is returned when DiffSchemas reports a
// non-empty delta and the caller (CommitTable handler) is operating
// in strict mode. Mirrors Rust's ApiError::SchemaIncompatible
// envelope so the 422 response shape stays byte-exact.
type SchemaIncompatibleError struct {
	CurrentSchema   json.RawMessage `json:"current_schema"`
	AttemptedSchema json.RawMessage `json:"attempted_schema"`
	Diff            SchemaDiff      `json:"diff"`
}

// Error implements `error` for SchemaIncompatibleError.
func (e *SchemaIncompatibleError) Error() string {
	return "schema strict-mode: writes diverge from current schema (" + e.Diff.Rendered() + ")"
}
